be/src/connector/mysql_connector.cpp

```diff
@@ -0,0 +1,463 @@
+    M(TYPE_CHAR, APPEND_TO_SQL)                       \
+    M(TYPE_VARCHAR, APPEND_TO_SQL)
+
+using namespace vectorized;
+
+// ================================
+
+DataSourceProviderPtr MySQLConnector::create_data_source_provider(vectorized::ConnectorScanNode* scan_node,
+                                                                  const TPlanNode& plan_node) const {
+    return std::make_unique<MySQLDataSourceProvider>(scan_node, plan_node);
+}
+
+// ================================
+
+MySQLDataSourceProvider::MySQLDataSourceProvider(vectorized::ConnectorScanNode* scan_node, const TPlanNode& plan_node)
+        : _scan_node(scan_node), _mysql_scan_node(plan_node.mysql_scan_node) {}
+
+DataSourcePtr MySQLDataSourceProvider::create_data_source(const TScanRange& scan_range) {
+    return std::make_unique<MySQLDataSource>(this, scan_range);
+}
+
+// ================================
+
+MySQLDataSource::MySQLDataSource(const MySQLDataSourceProvider* provider, const TScanRange& scan_range)
+        : _provider(provider) {}
+
+Status MySQLDataSource::_init_params(RuntimeState* state) {
+    VLOG(1) << "MySQLDataSource::init mysql scan params";
+
+    DCHECK(state != nullptr);
+
+    _columns = _provider->_mysql_scan_node.columns;
+    _filters = _provider->_mysql_scan_node.filters;
+
+    _table_name = _provider->_mysql_scan_node.table_name;
+
+    // get tuple desc
+    _tuple_desc = state->desc_tbl().get_tuple_descriptor(_provider->_mysql_scan_node.tuple_id);
+    DCHECK(_tuple_desc != nullptr);
+
+    _slot_num = _tuple_desc->slots().size();
+    // get mysql info
+    const auto* mysql_table = dynamic_cast<const MySQLTableDescriptor*>(_tuple_desc->table_desc());
+    DCHECK(mysql_table != nullptr);
+
+    _my_param.host = mysql_table->host();
+    _my_param.port = mysql_table->port();
+    _my_param.user = mysql_table->user();
+    _my_param.passwd = mysql_table->passwd();
+    _my_param.db = mysql_table->mysql_db();
+    // new one scanner
+    _mysql_scanner = std::make_unique<MysqlScanner>(_my_param);
+    DCHECK(_mysql_scanner != nullptr);
+
+    return Status::OK();
+}
+
+Status MySQLDataSource::open(RuntimeState* state) {
+    _init_params(state);
+    DCHECK(state != nullptr);
+    RETURN_IF_CANCELLED(state);
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+    RETURN_IF_ERROR(_mysql_scanner->open());
+
+    // Get [slot_id, slot] map
+    std::unordered_map<SlotId, SlotDescriptor*> slot_by_id;
+    for (SlotDescriptor* slot : _tuple_desc->slots()) {
+        slot_by_id[slot->id()] = slot;
+    }
+
+    std::unordered_map<std::string, std::vector<std::string>> filters_in;
+    std::unordered_map<std::string, bool> filters_null_in_set;
+
+    // In Filter have been put into _conjunct_ctxs,
+    // so we iterate all ExprContext to use it.
+    for (int i = 0; i < _conjunct_ctxs.size(); ++i) {
+        ExprContext* ctx = _conjunct_ctxs[i];
+        const Expr* root_expr = ctx->root();
+        if (root_expr != nullptr) {
```

Comment:
```suggestion
        if (root_expr == nullptr) {
            continue;
        }
     
```
make code line shorter